### PR TITLE
Chore: elevate how authentication is done

### PR DESCRIPTION
**Pull request by till**
##### SUMMARY

I spent a few days trying to figure out where bad requests came from when ansible ran and attempted to download public downloads from Github repositories. I discovered that a netrc configuration is used when in the environment and apparently the configuration contained within is added to requests.

For context, the netrc file is setup "by" our CI server (to be able to clone private github repositories), but I was unaware of this. I saw there are a few tickets here and on multiple Ansible roles where people stumble into this and I thought it was a good idea to add some logging to the file. But this is more or less an RFC, happy to change/adjust.

This code seems to affect a bunch of different modules.

Related: #61666 #27293

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

 - `uri`
 - `get_url`
 - `lookup('url', ...)`

### CODE REVIEW

**Comment by till** [comment]
@gundalow Is there a way to retrigger Shippable? It seemed like some unrelated issue why the tests didn't run.

File: lib/ansible/module_utils/urls.py [hunk 1]
```diff
@@ -1145,6 +1146,10 @@ def open(self, method, url, data=None, headers=None, use_proxy=None,
 
         method = method.upper()
 
+        logger = logging.getLogger(__name__)
+        logger.setLevel(logging.DEBUG)
+        logger.addHandler(logging.StreamHandler(sys.stdout))
```
Comment (bcoca) [inline lib/ansible/module_utils/urls.py]: We avoid creating loggers for each system, the module class already has it's own logger.

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/urls.py (hunk 1 above)
Comment (till) [reply lib/ansible/module_utils/urls.py]: @bcoca Can you point me to what you mean? I was looking for a logger, couldn't find one.

Comment (jborean93) [reply lib/ansible/module_utils/urls.py]: The `module` object has methods like `info()`, `debug()`, `warn()` that you can use. These details are sent to the syslogger defined by the ansible config.

Comment (till) [reply lib/ansible/module_utils/urls.py]: @jborean93 Thank you, I see what you mean. And is it "acceptable" to pass `module` to `def open(...)`?

Comment (till) [reply lib/ansible/module_utils/urls.py]: So, it seems like this is not easily doable. Also, probably a hefty BC break if I change the API of `open()`. Can anyone give me a hint how to use module there? The class `Request` is currently unaware of it. Can I inject it in `__init__` or what would you prefer?

Comment (till) [reply lib/ansible/module_utils/urls.py]: @jborean93 @bcoca Any feedback? Love to finish this and close this out.

Comment (sivel) [reply lib/ansible/module_utils/urls.py]: The code in urls.py has no access to the module object as you mention, nor should it have access as it's not required to run in a module context.  This puts us in a difficult place, as we cannot and should not have access to it, and just using `logging` is not correct.

We would likely need to abstract the syslog functionality out of the `AnsibleModule` class, similarly to what we did with warnings, and then allow `AnsibleModule` to consume or augment that functionality.  This is likely going to be a large undertaking.

As such, I don't see us adding this functionality right now.

File: lib/ansible/module_utils/urls.py [hunk 2]
```diff
@@ -1051,7 +1051,7 @@ class Request:
     def __init__(self, headers=None, use_proxy=True, force=False, timeout=10, validate_certs=True,
                  url_username=None, url_password=None, http_agent=None, force_basic_auth=False,
                  follow_redirects='urllib2', client_cert=None, client_key=None, cookies=None, unix_socket=None,
-                 ca_path=None):
+                 ca_path=None, module=None):
```
Comment (sivel) [inline lib/ansible/module_utils/urls.py:1054]: I still do not believe this should be here.  This class should not have any knowledge of modules.  I realize the alternative is insanely more complex, but that doesn't mean that this is the right choice due to ease.

File: lib/ansible/module_utils/urls.py (hunk 1 above)
Comment (till) [reply lib/ansible/module_utils/urls.py]: Do you think we can turn off netrc then?

This was a huge time sink, trying to figure out what happens so deep in internals.

Comment (sivel) [reply lib/ansible/module_utils/urls.py]: We may be able to add an option that can disable netrc, with the default still to use netrc.  I'm not sure that helps you though.  Although I'm not sure how disabling netrc at this point helps you.

File: lib/ansible/module_utils/urls.py (hunk 2 above)
Comment (till) [reply lib/ansible/module_utils/urls.py:1054]: Yeah, I agree the composition there is bad.

I like the idea of ripping out the logging. Haven't found yet where it all gets put together. As in... where the `AnsibleModule` is created.

**Comment by s-hertel** [comment]
Closing as per https://github.com/ansible/ansible/pull/45568#discussion_r227854878. We shouldn't be passing the module.

**Pull request closed without merging**